### PR TITLE
AP_Scripting: log memory cost of initial load and add debug print for state memory cost

### PR DESCRIPTION
**Pull request by IamPete1**
This give us a little more insight into where memory is being used.

### CODE REVIEW

File: libraries/AP_Scripting/lua_scripts.cpp [hunk 1]
```diff
@@ -427,9 +462,19 @@ void lua_scripts::run(void) {
...
+#ifndef HAL_CONSOLE_DISABLED
+    const int inital_mem = lua_gc(L, LUA_GCCOUNT, 0) * 1024 + lua_gc(L, LUA_GCCOUNTB, 0);
+#endif
+
     lua_atpanic(L, atpanic);
     load_generated_bindings(L);
 
+#ifndef HAL_CONSOLE_DISABLED
+    const int loaded_mem = lua_gc(L, LUA_GCCOUNT, 0) * 1024 + lua_gc(L, LUA_GCCOUNTB, 0);
+    DEV_PRINTF("Lua: State memory usage: %i + %i\n", inital_mem, loaded_mem - inital_mem);
```
Comment (IamPete1) [inline libraries/AP_Scripting/lua_scripts.cpp:475]: Could put this behind the runtime debug option also. Or a new define, it does not change per run.

**Comment by timtuxworth** [comment]
Could you not log it, i.e. as a telemetry sensor? Then it could be displayed in MP and graphed ...

**Comment by WickedShell** [comment]
It's important to note that the run memory stats here are all just how much garbage was left at the end of the loop, if you used a ton of memory and GC'd before that you won't see it reflected in these stats. Ideally we would track the number of times the GC was run.

File: libraries/AP_Scripting/lua_scripts.cpp [hunk 2]
```diff
@@ -117,6 +117,34 @@ int lua_scripts::atpanic(lua_State *L) {
     return 0;
 }
 
+// helper for print and log of runtime stats
+void lua_scripts::update_stats(const char *name, uint32_t run_time, int total_mem, int run_mem)
+{
+    if ((_debug_options.get() & uint8_t(DebugLevel::RUNTIME_MSG)) != 0) {
```
Comment (peterbarker) [inline libraries/AP_Scripting/lua_scripts.cpp:123]: Ewww.

Is it time for
```suggestion
    if (debug_option_is_set(DebugLevel::RUNTIME_MSG))) {
```

... also, is this a bitmask or a "level"?!

**Review by tridge — APPROVED** [review]
looks good, but the option cleanup @peterbarker suggests would be a nice bonus :-)

File: libraries/AP_Scripting/lua_scripts.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_Scripting/lua_scripts.cpp:123]: The problem is that `_debug_options` is a const reference to the parameter in the front end. It is used both here and in the front end. We could do a `debug_option_is_set` in the front end, but then we would have to pass a front end object in or go via the singleton.

So TLDR its more involved to add a helper than it seems. 

It was a level, we changed to a bitmask but did not rename everything.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**